File: python/src/curious/git_branch.py
from __future__ import annotations

import subprocess
from pathlib import Path

DEFAULT_AGENT_BRANCH = "curious"


def _run_git(cwd: Path, *args: str) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        ["git", *args],
        cwd=cwd,
        capture_output=True,
        text=True,
        check=False,
    )


def git_toplevel(start: Path) -> Path | None:
    result = _run_git(start, "rev-parse", "--show-toplevel")
    if result.returncode != 0:
        return None
    return Path(result.stdout.strip())


def is_git_repo(start: Path) -> bool:
    return git_toplevel(start) is not None


def current_branch(repo_root: Path) -> str | None:
    result = _run_git(repo_root, "rev-parse", "--abbrev-ref", "HEAD")
    if result.returncode != 0:
        return None
    branch = result.stdout.strip()
    return branch or None


def branch_exists(repo_root: Path, name: str) -> bool:
    result = _run_git(
        repo_root,
        "show-ref",
        "--verify",
        "--quiet",
        f"refs/heads/{name}",
    )
    return result.returncode == 0


def ensure_agent_branch(
    project_root: Path,
    branch: str = DEFAULT_AGENT_BRANCH,
    *,
    enabled: bool = True,
) -> str | None:
    """Switch the repo to the Curious agent branch. Returns branch name or None if skipped."""
    if not enabled:
        return None

    root = git_toplevel(Path(project_root))
    if root is None:
        print("[curious] git: not a repository — skipping agent branch switch")
        return None

    current = current_branch(root)
    if current == branch:
        print(f"[curious] git: already on branch {branch}")
        return branch

    if branch_exists(root, branch):
        result = _run_git(root, "switch", branch)
        if result.returncode != 0:
            result = _run_git(root, "checkout", branch)
    else:
        result = _run_git(root, "switch", "-c", branch)
        if result.returncode != 0:
            result = _run_git(root, "checkout", "-b", branch)

    if result.returncode != 0:
        detail = (result.stderr or result.stdout or "").strip()
        raise RuntimeError(
            f"Failed to switch to branch {branch!r}"
            + (f" (from {current})" if current else "")
            + (f": {detail}" if detail else "")
        )

    print(
        f"[curious] git: switched to branch {branch}"
        + (f" (from {current})" if current else "")
    )
    return branch


def agent_branch_prompt_note(branch: str) -> str:
    return (
        f"\n**Branch:** Curious switched this repo to `{branch}` before this run. "
        "Stay on this branch; do not run `git switch`, `git checkout`, or worktrees.\n"
    )

File: src/git-branch.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

export const DEFAULT_AGENT_BRANCH = "curious";

async function runGit(
  cwd: string,
  args: string[],
): Promise<{ code: number; stdout: string; stderr: string }> {
  try {
    const { stdout, stderr } = await execFileAsync("git", args, { cwd });
    return { code: 0, stdout: stdout.trim(), stderr: stderr.trim() };
  } catch (err: unknown) {
    const e = err as { code?: number; stdout?: string; stderr?: string };
    return {
      code: e.code ?? 1,
      stdout: (e.stdout ?? "").toString().trim(),
      stderr: (e.stderr ?? "").toString().trim(),
    };
  }
}

export async function gitToplevel(start: string): Promise<string | null> {
  const result = await runGit(start, ["rev-parse", "--show-toplevel"]);
  if (result.code !== 0) return null;
  return result.stdout || null;
}

export async function currentBranch(repoRoot: string): Promise<string | null> {
  const result = await runGit(repoRoot, ["rev-parse", "--abbrev-ref", "HEAD"]);
  if (result.code !== 0) return null;
  return result.stdout || null;
}

async function branchExists(repoRoot: string, name: string): Promise<boolean> {
  const result = await runGit(repoRoot, [
    "show-ref",
    "--verify",
    "--quiet",
    `refs/heads/${name}`,
  ]);
  return result.code === 0;
}

export function agentBranchPromptNote(branch: string): string {
  return (
    `\n**Branch:** Curious switched this repo to \`${branch}\` before this run. ` +
    "Stay on this branch; do not run `git switch`, `git checkout`, or worktrees.\n"
  );
}

/**
 * Switch the repo to the Curious agent branch before agent work.
 * Returns the branch name, or null if skipped (not a git repo / disabled).
 */
export async function ensureAgentBranch(
  projectRoot: string,
  branch = DEFAULT_AGENT_BRANCH,
  enabled = true,
): Promise<string | null> {
  if (!enabled) return null;

  const root = await gitToplevel(projectRoot);
  if (!root) {
    console.log("[curious] git: not a repository — skipping agent branch switch");
    return null;
  }

  const current = await currentBranch(root);
  if (current === branch) {
    console.log(`[curious] git: already on branch ${branch}`);
    return branch;
  }

  let result: { code: number; stderr: string; stdout: string };
  if (await branchExists(root, branch)) {
    result = await runGit(root, ["switch", branch]);
    if (result.code !== 0) {
      result = await runGit(root, ["checkout", branch]);
    }
  } else {
    result = await runGit(root, ["switch", "-c", branch]);
    if (result.code !== 0) {
      result = await runGit(root, ["checkout", "-b", branch]);
    }
  }

  if (result.code !== 0) {
    const detail = result.stderr || result.stdout;
    throw new Error(
      `Failed to switch to branch ${JSON.stringify(branch)}` +
        (current ? ` (from ${current})` : "") +
        (detail ? `: ${detail}` : ""),
    );
  }

  console.log(
    `[curious] git: switched to branch ${branch}` +
      (current ? ` (from ${current})` : ""),
  );
  return branch;
}

export async function agentBranchPromptNoteForConfig(args: {
  projectRoot: string;
  agentBranch?: string;
  ensureAgentBranch?: boolean;
}): Promise<string> {
  if (args.ensureAgentBranch === false) return "";
  const root = await gitToplevel(args.projectRoot);
  if (!root) return "";
  const branch = (await currentBranch(root)) ?? args.agentBranch ?? DEFAULT_AGENT_BRANCH;
  return agentBranchPromptNote(branch);
}
